### DISCUSSION TITLE
Can't figure out setting up Environment variables/Secrets?

### DISCUSSION THREAD

**Discussion by tettoffensive** [vercel/vercel#4558] [Help] [answered]
I'm having trouble understanding how environment variables and secrets work with vercel. I have a Next.js project I'm using with firebase. I want my firebase private key to be a secret accessible through an environment variable. While I've seen a few commands for setting up secrets, I'm missing how to connect everything up. I haven't found a good walkthrough.

Here's what I've done so far:

1. Initialize my project
```
vercel
```
2. Extract the private key from .json file downloaded from firebase and put into
**firebase-private.key**
```
-----BEGIN PRIVATE KEY-----
…
-----END PRIVATE KEY-----
```
3. Create my secret from a key file
```
vc secret add firebase-private-key "$(cat firebase-private.key | base64 | tr -d \\n)"
```
4. Ran this, though not sure it did anything. The environment variable doesn't seem to exist.
```shell
vc -e FIREBASE_PRIVATE_KEY=@firebase-private-key
```
5. I would've expected the last command to do the following, but I manually added:
**.vercel/project.json**
```json
{
  "projectId":"…",
  "orgId":"…",
  "env": {
    "FIREBASE_PRIVATE_KEY": "@firebase-private-key"
  }
}
```
6. I've been using a .env for local development, but I don't know how to link this up with what vercel has.
**.env**
```bash
FIREBASE_PROJECT_ID='my-project'
FIREBASE_PRIVATE_KEY='What goes here?'
```

- Why doesn't vercel's dashboard show any Environment variables?
- Can I populate vercel's environment variables with my .env?
- How can I keep .env & .vercel/project.json & `vercel env ls` in sync? Should I drop .env? Can .env read from vercel?

**Comment by mcsdevv** [comment, marked as answer]
Hey @tettoffensive, thanks for raising this!

The best way to handle this is entirely through the Vercel dashboard: https://vercel.com/docs/v2/build-step#environment-variables

There you can add all environment variables (encrypted by default) for all environments, Production, Preview, and Development.

During development, all you need to do is use `vercel env pull`, this will create (or update) a `.env` file you have locally, making the variables defined for Development available to your application.

There is no need to use `vercel secrets` anymore unless you require an account level environment variable.

**Reply by tettoffensive** [reply]
I see. So for my multiline `-----BEGIN PRIVATE KEY-----` should I set it a single line string with `\n` throughout it.

**Reply by ebartan** [reply]
> I see. So for my multiline `-----BEGIN PRIVATE KEY-----` should I set it a single line string with `\n` throughout it.

this solution works for me! thx mate

**Reply by stevez86** [reply]
Hi @mcsdevv Quick question; is there any way to create secrets that are obscured in the dashboard? It seems like a big security risk to not make them hidden somehow

**Reply by amyegan** [reply]
Hi @stevez86! Environment variables are obscured by default in the Dashboard. For example, this is a view of the environment variables in one my my test apps:

<img width="744" alt="image" src="https://user-images.githubusercontent.com/8247105/191025802-774df55f-2015-4450-a11c-848da9b0734e.png">

**Reply by Nicholas-Arthur-Cook** [reply]
Hi @amyegan . This still seems like a security risk, since the secret can be exposed by pressing the eye icon. Ideally secrets should not be able to be "recreated" like this, so that if there were a security breach, the damage would be limited. This means that if someone gained access to our Vercel account, they would have access to _all_ of our other associated services with API keys in secrets.

**Reply by amyegan** [reply]
Hi @Nicholas-Arthur-Cook. We do have some [RBAC roles](https://vercel.com/docs/teams-and-accounts/team-members-and-roles) which allow for restrictions on editing environment variables, however they are currently limited to Enterprise teams. At the moment, you can only assign the [Member](https://vercel.com/docs/teams-and-accounts/team-members-and-roles#member-role) and [Owner](https://vercel.com/docs/teams-and-accounts/team-members-and-roles#owner-role) roles on a Pro plan, and Hobby accounts are essentially owner by default. Both roles allow for access to all of the environment variables.

I shared your feedback with the Engineering team and made a note to share an update here for any env var security-related changes. You can also follow the [changelog](https://vercel.com/changelog) Vercel feature updates

**Comment by styfle** [comment]
If you want to use the CLI with environment variables, you can use `vercel env` subcommand (see `vercel env --help)`.

```sh
cat firebase-private-dev.key | vc env add FIREBASE_PRIVATE_KEY development
cat firebase-private-prev.key | vc env add FIREBASE_PRIVATE_KEY preview
cat firebase-private-prod.key | vc env add FIREBASE_PRIVATE_KEY production
```

Then you can use `vc env pull` to create a local `.env` file which will be used for `vc dev` or any other frontend framework.

Note: you do not need a `vercel.json` file at all, this will work without configuration files.

**Comment by leerob** [comment]
Hey folks, good news. We've made this much easier in two main ways:

1. You can now copy paste your `.env` file directly into the the input! ([source](https://twitter.com/leeerob/status/1593331366292250625))
2. When you paste in private key secrets, we'll automatically convert `\n` to new lines ([source](https://twitter.com/leeerob/status/1595089797164441600))

**Reply by NicHaley** [reply]
I was pasting a multiline GitHub env var and the command + v saved me 😌
